File: 2024-12-December-GFG-POTD/December 19 - Kth Missing Positive Number in a Sorted Array/Solution.py
class Solution:
    def kthMissing(self, arr, k):
        missing_count = 0  # Counts missing numbers
        current = 1        # Start from the smallest positive number
        index = 0          # Tracks position in the array

        while True:
            if index < len(arr) and arr[index] == current:
                # Current number is in the array, move to the next number
                index += 1
            else:
                # Current number is missing
                missing_count += 1
                if missing_count == k:
                    return current  # Found the kth missing number
            current += 1


if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        arr = list(map(int, input().split()))
        k = int(input().strip())

        sol = Solution()
        result = sol.kthMissing(arr, k)

        print(result)
        print("~")

File: 2024-12-December-GFG-POTD/December 19 - Kth Missing Positive Number in a Sorted Array/Solution.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

class Solution {
public:
    int kthMissing(vector<int>& arr, int k) {
        int missingCount = 0; // Counts missing numbers
        int current = 1;      // Start from the smallest positive number
        int index = 0;        // Tracks position in the array

        while (true) {
            if (index < arr.size() && arr[index] == current) {
                // Current number is in the array, move to the next number
                index++;
            } else {
                // Current number is missing
                missingCount++;
                if (missingCount == k) {
                    return current; // Found the kth missing number
                }
            }
            current++;
        }
    }
};

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        string line;
        getline(cin, line);
        stringstream ss(line);
        vector<int> arr;
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }

        int k;
        cin >> k;
        cin.ignore();

        Solution sol;
        int result = sol.kthMissing(arr, k);

        cout << result << endl;
        cout << "~" << endl;
    }

    return 0;
}

